src/viper/syntax.ml

```diff
@@ -58,6 +58,9 @@ and invariants = exp list
 
 and stmt = (stmt', info) Source.annotated_phrase
 
+and tmpl = (tmpl', info) Source.annotated_phrase
+and tmpl' = (Mo_def.Syntax.exp -> exp) -> exp
```

Comment:
Mixing the two syntaxes (Motoko and Viper) is kinda gross, but we can let it slide for now.

I wonder if this would be simpler if you just collected the async invariants from the motoko term during the translation and translated eager to vanilla viper, rather than introducing the intermediate, artificial ConcurrencyS statement. Ditto for PreconditionS and PostconditionS. None of these are mean to appear in the final target term, right?

Reply:
Yep, they get all filtered out, eventually. The problem is that I cannot pretty-print the Viper term as the `DefinitionI` base context is not available there; it gets created later on. Also `trans.ml` should not depend on `pretty.ml`, right?

Reply:
Actually, I don't mind `InvartiantI` since it closely corresponds to a MacroDefinition in Viper (source) syntax. So we could keep that but rename to `DefineI`. But we can sort this out later...